diff --git a/drivers/remoteproc/imx_rproc.c b/drivers/remoteproc/imx_rproc.c
index 6d913ab..6115623 100644
--- a/drivers/remoteproc/imx_rproc.c
+++ b/drivers/remoteproc/imx_rproc.c
@@ -411,7 +411,7 @@ static int imx_rproc_start(struct rproc *rproc)
 		}
 	} else if (priv->enable) {
 		if (!reset_control_status(priv->enable)) {
-			dev_info(dev, "alreay started\n");
+			dev_info(dev, "already started\n");
 			return 0;
 		}
 
@@ -464,7 +464,7 @@ static int imx_rproc_stop(struct rproc *rproc)
 		return -EBUSY;
 	}
 
-	if (priv->txdb_ch) {
+	if (!IS_ERR(priv->txdb_ch)) {
 		ret = mbox_send_message(priv->txdb_ch, (void *)&mmsg);
 		if (ret) {
 			dev_err(dev, "txdb send fail: %d\n", ret);
@@ -792,21 +792,31 @@ static void imx_rproc_kick(struct rproc *rproc, int vqid)
 	mmsg = vqid << 16;
 
 	priv->cl.tx_tout = 50;
-	err = mbox_send_message(priv->tx_ch, (void *)&mmsg);
-	if (err < 0)
-		dev_err(priv->dev, "%s: failed (%d, err:%d)\n",
-			__func__, vqid, err);
+	if(!IS_ERR(priv->tx_ch))
+	{
+		err = mbox_send_message(priv->tx_ch, (void *)&mmsg);
+		if (err < 0)
+			dev_err(priv->dev, "%s: failed (%d, err:%d)\n",
+				__func__, vqid, err);
+	}
 }
 
 static void *imx_rproc_memcpy(struct rproc *rproc, void *dest,
 			      const void *src, size_t count, int flags)
 {
-	u32 *tmp = dest;
-	const u32 *s = src;
+	union{u32 *p32;       u8 *p8;}tmp;
+	union{const u32 *p32; const u8 *p8;}s;
+	u32 countbytes;
 
-	count = count / 4;
+	tmp.p32 = dest;
+	s.p32 = src;
+
+	countbytes = count & (4-1);
+	count = count / 4u;
 	while (count--)
-		*tmp++ = *s++;
+		*tmp.p32++ = *s.p32++;
+	while (countbytes--)
+		*tmp.p8++ = *s.p8++;
 
 	return dest;
 }
@@ -984,7 +994,7 @@ static int imx_rproc_db_channel_init(struct rproc *rproc)
 	if (IS_ERR(priv->txdb_ch)) {
 		dev_info(cl->dev, "No txdb, ret %d\n",
 			ret);
-		priv->txdb_ch = NULL;
+		/*priv->txdb_ch = NULL;*/
 	}
 
 	return ret;
@@ -1136,10 +1146,18 @@ static int imx_rproc_probe(struct platform_device *pdev)
 
 	INIT_DELAYED_WORK(&(priv->rproc_work), imx_rproc_vq_work);
 
-	ret = imx_rproc_xtr_mbox_init(rproc);
-	if (ret) {
-		if (ret == -EPROBE_DEFER)
-			goto err_put_rproc;
+	if(of_property_read_bool(dev->of_node, "mailbox-names"))
+	{
+		ret = imx_rproc_xtr_mbox_init(rproc);
+		if (ret) {
+			if (ret == -EPROBE_DEFER)
+				goto err_put_rproc;
+		}
+	}
+	else
+	{
+		priv->rx_ch = ERR_PTR(-ENODEV);
+		priv->tx_ch = ERR_PTR(-ENODEV);
 	}
 
 	priv->clk = devm_clk_get_optional(dev, NULL);
@@ -1159,9 +1177,17 @@ static int imx_rproc_probe(struct platform_device *pdev)
 		goto err_put_rproc;
 	}
 
-	ret = imx_rproc_db_channel_init(rproc);
-	if (ret)
-		goto err_put_mbox;
+	if(of_property_read_bool(dev->of_node, "mailbox-names"))
+	{
+		ret = imx_rproc_db_channel_init(rproc);
+		if (ret)
+			goto err_put_mbox;
+	}
+	else
+	{
+		priv->rxdb_ch = ERR_PTR(-ENODEV);
+		priv->txdb_ch = ERR_PTR(-ENODEV);
+	}
 
 #ifdef CONFIG_IMX_SCU
 	priv->proc_nb.notifier_call = imx_rproc_partition_notify;
